### PR TITLE
receiver/prometheus: wire up direct pdata conversion mechanism

### PR DESCRIPTION
**Pull request by Aneurysm9**
**Description:** Adds the remaining pieces of a direct-to-pdata mechanism for the Prometheus receiver and a feature gate that allows selecting whether to use the legacy OpenCensus translations or the new direct route.

**Link to tracking Issue:** #4892

**Testing:** Unit tests exist for new code in the `internal` package.  End-to-end tests for the receiver now run against configurations that use both the OpenCensus and pdata builders.

### CODE REVIEW

**Comment by Aneurysm9** [comment]
@dashpole can you take a look at this?

**Comment by Aneurysm9** [comment]
Changes outside of `receiver/prometheusreceiver` are due to https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6305 adding a CI check using `porto` but not ensuring that it was up-to-date with HEAD.  If another PR that has to make the same changes to pass the lint check lands first then they should disappear from this PR.

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:277]: Should this be debug?

File: receiver/prometheusreceiver/internal/otlp_metricsbuilder.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/otlp_metricsbuilder.go:63]: Should we remove these new comments?

File: receiver/prometheusreceiver/internal/otlp_transaction.go
Comment (dashpole) [inline receiver/prometheusreceiver/internal/otlp_transaction.go:202]: Should we error log instead of panicing?

File: receiver/prometheusreceiver/config.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/config.go:61]: There is only a global feature gate registry, so it can't necessarily be used in tests.  The feature gate registry also has a lock, so we recommend that the feature status is checked once and stored if it is needed repeatedly.  There is no mechanism for operators to change feature gate statuses during program runtime, so it should be safe to read this once and re-use it.

File: receiver/prometheusreceiver/internal/otlp_metricfamily.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metricfamily.go:354]: I think that's probably the correct thing to do.

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:41]: I don't think so.  The GC interval is still set to two minutes.  That would need to be addressed separately.

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:277]: Seems reasonable.  I think this generally shouldn't happen, but isn't really actionable either and thus makes sense as a debug log.

**Review by dashpole — APPROVED** [review]
looks great!

**Comment by dashpole** [comment]
Huh, i'm sure why my approval checkmark isn't green.  I'm in the code owners entry for the prometheus receiver

**Comment by Aneurysm9** [comment]
> Huh, i'm sure why my approval checkmark isn't green. I'm in the code owners entry for the prometheus receiver

Not sure what's going on there.  It doesn't look like there are any files outside `/receiver/prometheus/` so you should be an owner for all files in the review.

In any event, this should have more eyes on it.  @anuraaga and @codeboten can I get a review?

**Comment by Aneurysm9** [comment]
@dashpole apparently to be an effective code owner for review approval purposes you must have write access to the repository:

> The people you choose as code owners must have write permissions for the repository.
- [About CODEOWNERS](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners)

Though that seems to be contradicted later in the same document:

> A CODEOWNERS file uses a pattern that follows most of the same rules used in gitignore files, with some exceptions. The pattern is followed by one or more GitHub usernames or team names using the standard `@username` or `@org/team-name` format. Users must have `read` access to the repository and teams must have explicit `write` access, even if the team's members already have access. 
- [The same document](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#codeowners-syntax) (emphasis in original)

v0v

File: receiver/prometheusreceiver/internal/metricsutil_pdata.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/metricsutil_pdata.go:56]: From what I can tell this is referring to line 53

File: receiver/prometheusreceiver/internal/metricsutil_pdata.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/metricsutil_pdata.go:44]: From what I understand, all prometheus data is cumulative including gauge, but am likely wrong.

Wondering if there is a reference for the mapping, I searched around and found this doc with a broken link to a design doc. Would help reviewing this PR if there's information about the mapping somewhere

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/semantic_conventions/openmetrics-guidelines.md

File: receiver/prometheusreceiver/internal/otlp_metricfamily.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/otlp_metricfamily.go:141]: I don't think I can find any usage of this (sorry if I missed a collapsed large diff)

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:80]: Maybe better to use a hashed object instead of string for map key

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:93]: Since `gc()` uses a lock to access `tsi`, I'd expect a lock here too, or otherwise atomic write. If it's already locked at a higher level, then a comment could help

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (anuraaga) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:80]: I think the cloudwatch exporter has a pattern using a Sorted object that could work

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:171]: If `get` can be called from multiple threads, I think this can cause gc to be called multiple times concurrently, I suspect that's not intended. Can use a comment if this case can't happen since a simple method like `get` we'd generally expect to not have any threading guarantees

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:176]: Notice above using `printf` and here using concatenation, I think most code in collector would use printf

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:216]: It's somewhat mysterious we're locking on `tsm` here but not accessing it (presumably we are accessing through `adjustMetric`). Maybe a comment to ensure it was intended

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (anuraaga) [inline receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:253]: Debug level probably

**Review by anuraaga — COMMENTED** [review]
Intense PR did a first lightweight pass. I definitely wasn't expecting this much logic to convert between two metrics formats 😮 

High level question, is this embedding functionality from cumulativetodelta? And if so, should it be?

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/cumulativetodeltaprocessor/internal/tracking/metric.go

Also, things like metrics GC feel like they'd already be in the OTel SDK. Curious if there's no way to leverage the OTel SDK to do this conversion instead of what may be a lot of similar concepts. I wouldn't imagine there to be a huge penalty in efficiency, but not sure. And also just a wild idea that came to mind, it may not be applicable at all.

File: receiver/prometheusreceiver/internal/metricsutil_pdata.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/metricsutil_pdata.go:56]: I think this comment is either a vestige of a prior pdata API or a misunderstanding by the original author.  The default aggregation temporality appears to be unspecified.  See https://github.com/open-telemetry/opentelemetry-collector/blob/main/model/pdata/metrics.go#L256-L261

File: receiver/prometheusreceiver/internal/metricsutil_pdata.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/metricsutil_pdata.go:44]: I'm not actually sure.  This is inherited test helper code that seems to be used to test adjustment of gauge histograms, which should never be encountered since the appender won't build them and the metric adjuster will short circuit when encountering a distribution with other than cumulative temporality.  I think this can be removed without impacting functionality or test coverage.

File: receiver/prometheusreceiver/internal/otlp_metricfamily.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metricfamily.go:141]: It is used at the top of `toDistributionPoint()` and `toSummaryPoint()`.  They seem to be in collapsed parts of the diff.

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:93]: The lock is taken in `*MetricsAdjusterPdata.AdjustMetrics()`, which is the first exported method encountered when walking up the call stack from all invocations of this method.  This is the same as the OpenCensus-based implementation in metrics_adjuster.go that has been in use by this receiver for a long time.  I expect that to be the case for most of the comments on this implementation as most of them are in repeated code.

A reworking of this implementation is probably in order, particularly given the issue noted by @dashpole regarding targets that are scraped less frequently than the fixed GC interval.  That should probably be handled in a separate issue.

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:171]: The time since the last GC is checked again inside `gc()` once a lock is obtained.  It may be triggered multiple times, but it will still only run once per interval.  That's what the comments in this method and in `gc()` are trying to convey.

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:253]: This is currently info in the OC metrics adjuster.  If anything, since this is a situation that should never occur I might want to make this a warning.

File: receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go
Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:80]: I'm not sure I follow.  The `pdata.AttributeMap` isn't comparable and can't be used as a map key.  Since the labels are user-provided and arbitrary I'm not sure we could create a comparable object other than a string from them.

Comment (anuraaga) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:80]: There's this `attribute.Distinct` type used here for something similar I think recommended by jmacd a while back

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/243d6423992a4a76db30eb254d715ccb3a23c9e4/internal/aws/metrics/metric_calculator.go#L90

I figure it has at least somewhat better performance than strings

Comment (Aneurysm9) [reply receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go:80]: That is from the Go SDK's `attribute` package, which is not in use here.  These are `pdata.AttributeMap`s which would require similar conversion to utilize `attribute.Distinct`.

**Review by vishiy — APPROVED** [review]

**Comment by Aneurysm9** [comment]
@open-telemetry/collector-contrib-approvers can I get some more eyes on this? We have further work that is held up behind getting these changes landed.  Thanks!

**Comment by alolita** [comment]
Given @Aneurysm9 and @dashpole are codeowners for this component, can we get this PR merged please? Can the permissions be fixed @bogdandrutu @codeboten 

There are other PRs (14 PRs - 9 tests and 5 bug fixes) that are blocked on this merge.

**Comment by bogdandrutu** [comment]
@alolita I am not OK to merge this since a very good engineer and approver reviewed this (@anuraaga) and comments are not addressed.

**Review by anuraaga — APPROVED** [review]
I hadn't realized at first that there isn't that much new code here, a lot is copied from the old translation path to allow a feature flag deprecation. So generally good with the code as it seems to be a relatively simple change on top of that - would appreciate some more code comments to verify lock expectations though

**Comment by CatherineF-dev** [comment]
Hi, thanks for the great work which possibly reduce open-telemetry memory to 50% in a testing cluster. (v0.41 v.s. v0.40)

Want to cherry-pick this magic into previous versions. Do we know which small change in this PR is related to reduce agent memory?

**Comment by dashpole** [comment]
@CatherineF-dev I'm afraid this isn't an easily cherry-pickable change, as it involved refactoring the entire receiver.

**Pull request merged**
